workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/ViewRegistry.kt

```diff
@@ -102,10 +104,70 @@ public fun ViewRegistry(vararg bindings: Entry<*>): ViewRegistry =
 @WorkflowUiExperimentalApi
 public fun ViewRegistry(): ViewRegistry = TypedViewRegistry()
 
+/**
+ *  @throws IllegalArgumentException if the receiver already has a matching [entry].
+ */
+@WorkflowUiExperimentalApi
+public operator fun ViewRegistry.plus(entry: Entry<*>): ViewRegistry =
+  this + ViewRegistry(entry)
+
+/** @throws IllegalArgumentException if other has redundant entries. */
+@WorkflowUiExperimentalApi
+public operator fun ViewRegistry.plus(other: ViewRegistry): ViewRegistry {
+  if (other.keys.isEmpty()) return this
+  if (this.keys.isEmpty()) return other
+  return CompositeViewRegistry(this, other)
+}
+
+/**
+ * Replaces the existing [ViewRegistry] of the receiver with [registry]. Use
+ * [ViewEnvironment.merge] to combine them instead.
+ */
+@WorkflowUiExperimentalApi
+public operator fun ViewEnvironment.plus(registry: ViewRegistry): ViewEnvironment {
+  if (registry.keys.isEmpty()) return this
+  return this + (ViewRegistry to registry)
+}
+
+/**
+ * Combines the receiver with [other]. If there are conflicting entries,
+ * those in [other] are preferred.
+ */
+@WorkflowUiExperimentalApi
+public infix fun ViewRegistry.merge(other: ViewRegistry): ViewRegistry {
+  if (other.keys.isEmpty()) return this
+  if (this.keys.isEmpty()) return other
+
+  return (keys + other.keys).asSequence()
+    .map { other.getEntryFor(it) ?: getEntryFor(it)!! }
+    .toList()
+    .toTypedArray()
+    .let { ViewRegistry(*it) }
+}
+
+/**
+ * Merges the [ViewRegistry] of the receiver with [registry]. If there are conflicting entries,
+ * those in [registry] are preferred.
+ */
 @WorkflowUiExperimentalApi
-public operator fun ViewRegistry.plus(binding: Entry<*>): ViewRegistry =
-  this + ViewRegistry(binding)
+public infix fun ViewEnvironment.merge(registry: ViewRegistry): ViewEnvironment {
```

Comment:
Beware I think once we start having mixed typed operators things can sometimes get confusing - this is the case for `CoroutineContext` which maps scope, dispatcher, name etc. in a very similar way to this. 90% of the time things just 'work' and work in a quick and satisfying way, but sometimes when reasoning about it it can get confusing how things overlap.